# Données de 18/02/2019 00:00:15 jusqu'à 24/02/2019 20:50:57 des stations : 
- M7i (A7, Sens2 , $264+687$).
- M7I (A7, Sens1 , $264+687$).
- M7j (A7, Sens2 , $265+320$).
- M7J (A7, Sens1 , $265+320$).

In [1]:
#Charger les biblio
import numpy as np
import pandas as pd
from pandas import read_csv
import matplotlib.pyplot as plt
from matplotlib import cm
import seaborn as sns
import numpy as np
import string 
from subprocess import check_output
import seaborn as sns
from pylab import *
%matplotlib inline
sns.set()

In [2]:
# Lire la table avec 'header'
df = pd.read_csv('Hmvl33.csv', sep=';', engine='python')

In [3]:
# Remplacer les NaN par des zeros dans trame_hmvl
df.trame_hmvl.fillna('0', inplace=True)

In [4]:
# Reindexer la dataframe
df.index = range(len(df))

In [5]:
#changeaons les noms de colonnes
df.columns = ['code', 'id','equipement_id', 'date','trame']

In [6]:
# On enlève les @
df["trame"] = df["trame"].str.replace("@", "")

In [9]:
grouped = df.groupby('code')

In [27]:
M7i = grouped.get_group('M7i')
M7I = grouped.get_group('M7I')
M7j = grouped.get_group('M7j')
M7J = grouped.get_group('M7J')
liste = [M7i, M7J, M7j, M7I]

In [28]:
# Concatener les data
df = pd.concat(liste)

In [29]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
# Ajouter les colonnes, en séparant la colonne 'trame' 
for i in range(len(liste)):
    df[liste[i]] = df["trame"].str.slice(11*i,11*(i+1))

In [30]:
# Supprimer les colonnes xj
df0 = df.drop(['x3','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# df1
df1 = df.drop(['x3','x0','x2','x4','x5','x6','x7','x8', 'x9', 'x10',  'trame'], axis=1)
df2 = df.drop(['x3','x1','x0','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df3 = df.drop(['x0','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df4 = df.drop(['x0','x1','x2','x3','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df5 = df.drop(['x0','x1','x2','x4','x3','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df6 = df.drop(['x0','x1','x2','x4','x5','x3','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df7 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x8', 'x9', 'x10', 'trame'], axis=1)
df8 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x10', 'trame'], axis=1)
df9 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x8', 'x10',  'trame'], axis=1)

df10 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x8','trame'], axis=1)

In [31]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
liste2 = [df0,df1,df2,df3,df4,df5,df6,df7,df8,df9,df10]

    # Séparer la colonne xi en 4 colonnes  
for i, f in zip(range( len(liste)), liste2):
    f["voie"] = f[liste[i]].str.slice(0,1)
    f["horodatage"] = f[liste[i]].str.slice(1,5)
    f["vitesse"] = f[liste[i]].str.slice(5,8)
    f["longueur"] = f[liste[i]].str.slice(8,11)

In [32]:
# rename columns
for i,f in zip(range( len(liste)), liste2):
    
    f.rename(columns={liste[i]:'x'}, inplace=True)

In [33]:
# Concatener les data
result = pd.concat(liste2)

In [37]:
# On ajoute une colonne Count pour compter chacune
result['len'] = result['x'].str.len()
# Eliminer les lignes = 0 et 1
result = result[result.len != 0]
result = result[result.len != 1]

In [39]:
# Reindexer la dataframe
result.index = range(len(result))
result.head()

code    id  equipement_id                     date            x voie  \
0  M7i  6460           6460  2019-02-18 00:00:15.049  14148093033    1   
1  M7i  6460           6460  2019-02-18 00:00:26.549  05354096031    0   
2  M7i  6460           6460  2019-02-18 00:00:38.549  10050094039    1   
3  M7i  6460           6460  2019-02-18 00:00:50.548  01469102037    0   
4  M7i  6460           6460  2019-02-18 00:00:56.549  11973117040    1   

  horodatage vitesse longueur  len  
0       4148     093      033   11  
1       5354     096      031   11  
2       0050     094      039   11  
3       1469     102      037   11  
4       1973     117      040   11

In [40]:
# Créer une fonction pourcentage
def prc(part, whole):
    return 100 * float(part)/float(whole)

In [42]:
# Voyons le pourcentage des lignes dont 'len' diffenrent de 11
m = result[(result['len'] != 11)]
prc(len(m),len(result))

0.0

In [77]:
df = result.drop(['equipement_id','len', 'x'], axis=1)

In [78]:
#Exporter la table
df.to_csv('M7_i_j.csv', index = False, encoding = 'utf-8')

In [76]:
df.head()

code    id                     date            x voie horodatage vitesse  \
0  M7i  6460  2019-02-18 00:00:15.049  14148093033    1       4148     093   
1  M7i  6460  2019-02-18 00:00:26.549  05354096031    0       5354     096   
2  M7i  6460  2019-02-18 00:00:38.549  10050094039    1       0050     094   
3  M7i  6460  2019-02-18 00:00:50.548  01469102037    0       1469     102   
4  M7i  6460  2019-02-18 00:00:56.549  11973117040    1       1973     117   

  longueur  len  
0      033   11  
1      031   11  
2      039   11  
3      037   11  
4      040   11

In [75]:
grouped = df.groupby('id')
for name,group in grouped:
    print (name)


6431
6432
6460
6461


- M7i (A7, Sens2 , $264+687$) - id : 6460.
- M7I (A7, Sens1 , $264+687$) - id : 6431.
- M7j (A7, Sens2 , $265+320$) - id : 6461.
- M7J (A7, Sens1 , $265+320$) - id : 6432.

### Rassembler les données par tranche de 6mn


In [90]:
df['vitesse'] = pd.to_numeric(df['vitesse'] ,errors='coerce')

In [79]:
# library datetime
from datetime import datetime
from pandas import datetime

In [91]:
grouped = df.groupby('code')
M7i = grouped.get_group('M7i')
M7I = grouped.get_group('M7I')
M7j = grouped.get_group('M7j')
M7J = grouped.get_group('M7J')

In [ ]:
# Ou bien
l2 = ['M7i', 'M7J', 'M7j', 'M7I']
l1 = [M7i, M7J, M7j, M7I]
for i,f in zip(l1, range(len(l2))):
    i = grouped.get_group.(l2[f])

In [175]:
# convertir to datetime
df['date'] = pd.to_datetime(df['date'])
# Set the datetime column as the index
df.index = df['date']

In [194]:
d = df.groupby(['code','voie']).resample('6T')['vitesse'].mean()

In [195]:
type(d)

pandas.core.series.Series

In [196]:
# Conevertir la serie 'd' to dataframe
d = pd.DataFrame(d)

In [197]:
d.head()

vitesse
code voie date                           
M7I  0    2019-02-18 00:00:00   95.421053
          2019-02-18 00:06:00  103.857143
          2019-02-18 00:12:00  103.000000
          2019-02-18 00:18:00  101.200000
          2019-02-18 00:24:00  104.571429

In [198]:
d['date'] = d.index

In [199]:
# Reindexer la table
d.index = range(len(d))

In [200]:
d.head()

vitesse                           date
0   95.421053  (M7I, 0, 2019-02-18 00:00:00)
1  103.857143  (M7I, 0, 2019-02-18 00:06:00)
2  103.000000  (M7I, 0, 2019-02-18 00:12:00)
3  101.200000  (M7I, 0, 2019-02-18 00:18:00)
4  104.571429  (M7I, 0, 2019-02-18 00:24:00)

In [145]:
type(d.date)

pandas.core.series.Series

In [201]:
d["code"] = d['date'].str.slice(0,1)
d["voie"] = d['date'].str.slice(1,2)
d["date"] = d['date'].str.slice(2,3)

In [202]:
d.head(2)

vitesse                    date    code  voie
0   95.421053  (2019-02-18 00:00:00,)  (M7I,)  (0,)
1  103.857143  (2019-02-18 00:06:00,)  (M7I,)  (0,)

In [171]:
#d['date'] = pd.to_datetime(d['date'])

In [154]:
d["time"] = d['date'].str.slice(0,1)

In [155]:
d.head()

vitesse                    date    code                    time
0   99.800000  (2019-02-18 00:00:00,)  (M7I,)  (2019-02-18 00:00:00,)
1  100.903226  (2019-02-18 00:06:00,)  (M7I,)  (2019-02-18 00:06:00,)
2  105.363636  (2019-02-18 00:12:00,)  (M7I,)  (2019-02-18 00:12:00,)
3  101.485714  (2019-02-18 00:18:00,)  (M7I,)  (2019-02-18 00:18:00,)
4  101.363636  (2019-02-18 00:24:00,)  (M7I,)  (2019-02-18 00:24:00,)

essayons de le faire autrement

In [207]:
df1 = df.groupby(['code','voie']).resample('6T')['vitesse'].mean()

In [230]:
df1.index.names

FrozenList(['code', 'voie', 'date'])

In [227]:
indexe = df1.index.names

In [228]:
df = pd.DataFrame(df1, index = range(len(df1)), columns=indexe)